src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -851,28 +863,40 @@ public void AddAnalyzerReference(string fullPath)
         {
             CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
 
-            var visualStudioAnalyzer = new VisualStudioAnalyzer(
-                fullPath,
-                _hostDiagnosticUpdateSource,
-                Id,
-                Language);
-
             using (_gate.DisposableWait())
             {
                 if (_analyzerPathsToAnalyzers.ContainsKey(fullPath))
                 {
                     throw new ArgumentException($"'{fullPath}' has already been added to this project.", nameof(fullPath));
                 }
 
-                _analyzerPathsToAnalyzers.Add(fullPath, visualStudioAnalyzer);
-
-                if (_activeBatchScopes > 0)
+                // Are we adding one we just recently removed? If so, we can just keep using that one, and avoid removing
+                // it once we apply the batch
+                var analyzerPendingRemoval = _analyzersRemovedInBatch.FirstOrDefault(a => a.FullPath == fullPath);
+                if (analyzerPendingRemoval != null)
                 {
-                    _analyzersAddedInBatch.Add(visualStudioAnalyzer);
+                    _analyzersRemovedInBatch.Remove(analyzerPendingRemoval);
+                    _analyzerPathsToAnalyzers.Add(fullPath, analyzerPendingRemoval);
```

Comment:
Do we remove the path before the batch is complete? That seems odd... but maybe correct.

Reply:
Yeah, it's a bit strange. The _analyzerPathsToAnalyzers is added/removed to represent the "current" state of the world at all points; this means that any further add/remove can first check it for it being a valid action, i.e. the throw you see happening back a few lines before all of this if you're adding a duplicate.